presto-native-execution/presto_cpp/main/types/PrestoToVeloxQueryPlan.cpp

```diff
@@ -2151,23 +2151,24 @@ velox::core::PlanFragment VeloxQueryPlanConverter::toBatchVeloxQueryPlan(
     std::shared_ptr<std::string>&& serializedShuffleWriteInfo) {
   auto planFragment = toVeloxQueryPlan(fragment, tableWriteInfo, taskId);
 
-  // If the last node is a PartitionedOutputNode, it means this fragment ends
-  // with a shuffle stage. We convert the PartitionedOutputNode to a chain of
-  // following nodes:
+  // If the serializedShuffleWriteInfo is not nullptr, it means this fragment
+  // ends with a shuffle stage. We convert the PartitionedOutputNode to a
+  // chain of following nodes:
   // (1) A PartitionAndSerializeNode.
   // (2) A "gather" LocalPartitionNode that gathers results from multiple
   //     threads to one thread.
   // (3) A ShuffleWriteNode.
+  // To be noted, whether the last node of the plan is PartitionedOutputNode
+  // can't guarantee the query has shuffle stage, for example a plan with
+  // TableWriteNode can also havePartitionedOutputNode to distribute the
+  // metadata to coordinator.
+  if (serializedShuffleWriteInfo == nullptr) {
+    return planFragment;
+  }
   auto partitionedOutputNode =
       std::dynamic_pointer_cast<const core::PartitionedOutputNode>(
           planFragment.planNode);
-  VELOX_CHECK_EQ(
-      partitionedOutputNode == nullptr,
```

Comment:
Ah okay so this check is actually wrong. Good to know table write ends with an partitioned output node.